# Activity: Perform feature engineering 

## **Introduction**


As you're learning, data professionals working on modeling projects use featuring engineering to help them determine which attributes in the data can best predict certain measures.

In this activity, you are working for a firm that provides insights to the National Basketball Association (NBA), a professional North American basketball league. You will help NBA managers and coaches identify which players are most likely to thrive in the high-pressure environment of professional basketball and help the team be successful over time.

To do this, you will analyze a subset of data that contains information about NBA players and their performance records. You will conduct feature engineering to determine which features will most effectively predict whether a player's NBA career will last at least five years. The insights gained then will be used in the next stage of the project: building the predictive model.


## **Step 1: Imports** 


Start by importing `pandas`.

In [1]:
# Import pandas.

### YOUR CODE HERE ###
import pandas as pd

The dataset is a .csv file named `nba-players.csv`. It consists of performance records for a subset of NBA players. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# RUN THIS CELL TO IMPORT YOUR DATA.

# Save in a variable named `data`.

### YOUR CODE HERE ###

data = pd.read_csv("nba-players.csv", index_col=0)

<details><summary><h4><strong>Hint 1</strong></h4></summary>

The `read_csv()` function from `pandas` allows you to read in data from a csv file and load it into a DataFrame.
    
</details>

<details><summary><h4><strong>Hint 2</strong></h4></summary>

Call the `read_csv()`, pass in the name of the csv file as a string, followed by `index_col=0` to use the first column from the csv as the index in the DataFrame.
    
</details>

## **Step 2: Data exploration** 

Display the first 10 rows of the data to get a sense of what it entails.

In [3]:
# Display first 10 rows of data.

### YOUR CODE HERE ###
data.head(10)

name  gp   min  pts  fgm  fga    fg  3p_made  3pa    3p  ...  \
0   Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1  Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2   JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3      Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4      Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   
5     Tony Bennett  75  11.4  3.7  1.5  3.5  42.3      0.3  1.1  32.5  ...   
6      Don MacLean  62  10.9  6.6  2.5  5.8  43.5      0.0  0.1  50.0  ...   
7     Tracy Murray  48  10.3  5.7  2.3  5.4  41.5      0.4  1.5  30.0  ...   
8     Duane Cooper  65   9.9  2.4  1.0  2.4  39.2      0.1  0.5  23.3  ...   
9     Dave Johnson  42   8.5  3.7  1.4  3.5  38.3      0.1  0.3  21.4  ...   

   fta    ft  oreb  dreb  reb  ast  stl  blk  tov  target_5yrs  
0  2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3            0  
1  3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6            0  
2  1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0            0  
3  1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0            1  
4  1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8            1  
5  0.5  73.2   0.2   0.7  0.8  1.8  0.4  0.0  0.7            0  
6  1.8  81.1   0.5   1.4  2.0  0.6  0.2  0.1  0.7            1  
7  0.8  87.5   0.8   0.9  1.7  0.2  0.2  0.1  0.7            1  
8  0.5  71.4   0.2   0.6  0.8  2.3  0.3  0.0  1.1            0  
9  1.4  67.8   0.4   0.7  1.1  0.3  0.2  0.0  0.7            0  

[10 rows x 21 columns]

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

There is a function in the `pandas` library that can be called on a DataFrame to display the first n number of rows, where n is a number of your choice. 
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Call the `head()` function and pass in 10.
</details>

Display the number of rows and the number of columns to get a sense of how much data is available to you.

In [4]:
# Display number of rows, number of columns.

### YOUR CODE HERE ###
data.shape

(1340, 21)

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

DataFrames in `pandas` have an attribute that can be called to get the number of rows and columns as a tuple.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

You can call the `shape` attribute.
</details>

**Question:** What do you observe about the number of rows and the number of columns in the data?

There are 1340 rows and 21 columns.

Now, display all column names to get a sense of the kinds of metadata available about each player. Use the columns property in pandas.


In [6]:
# Display all column names.

### YOUR CODE HERE ###
#data.columns
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1340 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-null   float64
 19  tov   

The following table provides a description of the data in each column. This metadata comes from the data source, which is listed in the references section of this lab.

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played per game|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

Next, display a summary of the data to get additional information about the DataFrame, including the types of data in the columns.

In [7]:
# Use .info() to display a summary of the DataFrame.

### YOUR CODE HERE ###
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1340 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-null   float64
 19  tov   

**Question:** Based on the preceding tables, which columns are numerical and which columns are categorical?

All columns are numerical except 'name'. Games played ('gp') is a count integer. The 'target_5yrs' response variable is an integer representing a binary categorical variable. 

### Check for missing values

Now, review the data to determine whether it contains any missing values. Begin by displaying the number of missing values in each column. After that, use isna() to check whether each value in the data is missing. Finally, use sum() to aggregate the number of missing values per column.


In [10]:
# Display the number of missing values in each column.
# Check whether each value is missing.
#Aggregate the number of missing values per column.

### YOUR CODE HERE ###
data.isna().sum()

name           0
gp             0
min            0
pts            0
fgm            0
fga            0
fg             0
3p_made        0
3pa            0
3p             0
ftm            0
fta            0
ft             0
oreb           0
dreb           0
reb            0
ast            0
stl            0
blk            0
tov            0
target_5yrs    0
dtype: int64

**Question:** What do you observe about the missing values in the columns? 

There are no missing values in any of the columns.

**Question:** Why is it important to check for missing values?

Missing values can affect model training and performance, especially if the missing data are biased or systematically different from non-null data.

## **Step 3: Statistical tests** 



Next, use a statistical technique to check the class balance in the data. To understand how balanced the dataset is in terms of class, display the percentage of values that belong to each class in the target column. In this context, class 1 indicates an NBA career duration of at least five years, while class 0 indicates an NBA career duration of less than five years.

In [12]:
# Display percentage (%) of values for each class (1, 0) represented in the target column of this dataset.

### YOUR CODE HERE ###
data['target_5yrs'].value_counts(normalize=True)

1    0.620149
0    0.379851
Name: target_5yrs, dtype: float64

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

In `pandas`, `value_counts(normalize=True)` can be used to calculate the frequency of each distinct value in a specific column of a DataFrame.  
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

After `value_counts(normalize=True)`, multipling by `100` converts the frequencies into percentages (%).
</details>

**Question:** What do you observe about the class balance in the target column?

The target variable is slightly imbalanced (~62% 1's and ~38% 0's).

**Question:** Why is it important to check class balance?

Class imbalance can affect model training and performance.

## **Step 4: Results and evaluation** 


Now, perform feature engineering, with the goal of identifying and creating features that will serve as useful predictors for the target variable, `target_5yrs`. 

### Feature selection

The following table contains descriptions of the data in each column:

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

**Question:** Which columns would you select and avoid selecting as features, and why? Keep in mind the goal is to identify features that will serve as useful predictors for the target variable, `target_5yrs`. 

I might initially try including all predictor variables except ftm, 3p_made, and ftm, as this information will be captured by the combination of the two other corresponding variables (e.g., fga and fg capture fgm). Additionally, the 'reb' column incorporates both 'oreb' and 'dreb', and it is unlikely that a player with a high 'oreb' has a low 'dreb' or vice versa: I would expect that a good rebounder is typically good at both offensive and defensive rebounding. Finally, the name column also has no value for predicting target_5yrs.

Next, select the columns you want to proceed with. Make sure to include the target column, `target_5yrs`. Display the first few rows to confirm they are as expected.

In [34]:
# Select the columns to proceed with and save the DataFrame in new variable `selected_data`.
# Include the target column, `target_5yrs`.
### YOUR CODE HERE ###
selected_data = data.drop(['fgm', 'ftm', '3p_made', 'name', 'oreb', 'dreb'], axis=1)

# Display the first few rows.
### YOUR CODE HERE ###
selected_data.head()


gp   min  pts  fga    fg  3pa    3p  fta    ft  reb  ast  stl  blk  tov  \
0  36  27.4  7.4  7.6  34.7  2.1  25.0  2.3  69.9  4.1  1.9  0.4  0.4  1.3   
1  35  26.9  7.2  6.7  29.6  2.8  23.5  3.4  76.5  2.4  3.7  1.1  0.5  1.6   
2  74  15.3  5.2  4.7  42.2  1.7  24.4  1.3  67.0  2.2  1.0  0.5  0.3  1.0   
3  58  11.6  5.7  5.5  42.6  0.5  22.6  1.3  68.9  1.9  0.8  0.6  0.1  1.0   
4  48  11.5  4.5  3.0  52.4  0.1   0.0  1.9  67.4  2.5  0.3  0.3  0.4  0.8   

   target_5yrs  
0            0  
1            0  
2            0  
3            1  
4            1

In [35]:
selected_data.describe()

gp          min          pts          fga           fg  \
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000   
mean     60.414179    17.624627     6.801493     5.885299    44.169403   
std      17.433992     8.307964     4.357545     3.593488     6.137679   
min      11.000000     3.100000     0.700000     0.800000    23.800000   
25%      47.000000    10.875000     3.700000     3.300000    40.200000   
50%      63.000000    16.100000     5.550000     4.800000    44.100000   
75%      77.000000    22.900000     8.800000     7.500000    47.900000   
max      82.000000    40.900000    28.200000    19.800000    73.700000   

               3pa           3p          fta           ft          reb  \
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000   
mean      0.779179    19.149627     1.821940    70.300299     3.034478   
std       1.061847    16.051861     1.322984    10.578479     2.057774   
min       0.000000     0.000000     0.000000     0.000000     0.300000   
25%       0.000000     0.000000     0.900000    64.700000     1.500000   
50%       0.300000    22.200000     1.500000    71.250000     2.500000   
75%       1.200000    32.500000     2.300000    77.600000     4.000000   
max       6.500000   100.000000    10.200000   100.000000    13.900000   

               ast          stl          blk          tov  target_5yrs  
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000  
mean      1.550522     0.618507     0.368582     1.193582     0.620149  
std       1.471169     0.409759     0.429049     0.722541     0.485531  
min       0.000000     0.000000     0.000000     0.100000     0.000000  
25%       0.600000     0.300000     0.100000     0.700000     0.000000  
50%       1.100000     0.500000     0.200000     1.000000     1.000000  
75%       2.000000     0.800000     0.500000     1.500000     1.000000  
max      10.600000     2.500000     3.900000     4.400000     1.000000

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the materials about feature selection and selecting a subset of a DataFrame.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Use two pairs of square brackets, and place the names of the columns you want to select inside the innermost brackets. 

</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

There is a function in `pandas` that can be used to display the first few rows of a DataFrame. Make sure to specify the column names with spelling that matches what's in the data. Use quotes to represent each column name as a string. 
</details>

### Feature transformation

An important aspect of feature transformation is feature encoding. If there are categorical columns that you would want to use as features, those columns should be transformed to be numerical. This technique is also known as feature encoding.

**Question:** Why is feature transformation important to consider? Are there any transformations necessary for the features you want to use?

Feature transformation can improve model performance by ensuring that features fit model assumptions (e.g., normally distributed data) and can make certain features more interpretable for the modeling technique being used. In this case, there are no categorical predictor variables that need to be encoded.

### Feature extraction

Display the first few rows containing containing descriptions of the data for reference. The table is as follows:

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played per game|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

In [36]:
# Display the first few rows of `selected_data` for reference.
### YOUR CODE HERE ###
selected_data.head()


gp   min  pts  fga    fg  3pa    3p  fta    ft  reb  ast  stl  blk  tov  \
0  36  27.4  7.4  7.6  34.7  2.1  25.0  2.3  69.9  4.1  1.9  0.4  0.4  1.3   
1  35  26.9  7.2  6.7  29.6  2.8  23.5  3.4  76.5  2.4  3.7  1.1  0.5  1.6   
2  74  15.3  5.2  4.7  42.2  1.7  24.4  1.3  67.0  2.2  1.0  0.5  0.3  1.0   
3  58  11.6  5.7  5.5  42.6  0.5  22.6  1.3  68.9  1.9  0.8  0.6  0.1  1.0   
4  48  11.5  4.5  3.0  52.4  0.1   0.0  1.9  67.4  2.5  0.3  0.3  0.4  0.8   

   target_5yrs  
0            0  
1            0  
2            0  
3            1  
4            1

**Question:** Which columns lend themselves to feature extraction?

The 'min' feature in particular could be combined with many of the other features to express player efficiency in each of the categories (e.g., how many points per minute a player scores).

Extract two features that you think would help predict `target_5yrs`. Then, create a new variable named 'extracted_data' that contains features from 'selected_data', as well as the features being extracted.

In [37]:
# Extract two features that would help predict target_5yrs.
# Create a new variable named `extracted_data`.
### YOUR CODE HERE ### 
extracted_data = selected_data.join(pd.DataFrame({'pts_per_min':selected_data['pts'] / selected_data['min'], 
                                      'ast_per_min':selected_data['ast'] / selected_data['min'],
                                     'reb_per_min':selected_data['reb'] / selected_data['min'],
                                     'blk_per_min':selected_data['blk'] / selected_data['min'],
                                     'stl_per_min':selected_data['stl'] / selected_data['min'],
                                     'tov_per_min':selected_data['tov'] / selected_data['min']}))
extracted_data.head()

gp   min  pts  fga    fg  3pa    3p  fta    ft  reb  ...  stl  blk  tov  \
0  36  27.4  7.4  7.6  34.7  2.1  25.0  2.3  69.9  4.1  ...  0.4  0.4  1.3   
1  35  26.9  7.2  6.7  29.6  2.8  23.5  3.4  76.5  2.4  ...  1.1  0.5  1.6   
2  74  15.3  5.2  4.7  42.2  1.7  24.4  1.3  67.0  2.2  ...  0.5  0.3  1.0   
3  58  11.6  5.7  5.5  42.6  0.5  22.6  1.3  68.9  1.9  ...  0.6  0.1  1.0   
4  48  11.5  4.5  3.0  52.4  0.1   0.0  1.9  67.4  2.5  ...  0.3  0.4  0.8   

   target_5yrs  pts_per_min  ast_per_min  reb_per_min  blk_per_min  \
0            0     0.270073     0.069343     0.149635     0.014599   
1            0     0.267658     0.137546     0.089219     0.018587   
2            0     0.339869     0.065359     0.143791     0.019608   
3            1     0.491379     0.068966     0.163793     0.008621   
4            1     0.391304     0.026087     0.217391     0.034783   

   stl_per_min  tov_per_min  
0     0.014599     0.047445  
1     0.040892     0.059480  
2     0.032680     0.065359  
3     0.051724     0.086207  
4     0.026087     0.069565  

[5 rows x 21 columns]

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the materials about feature extraction.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Use the function `copy()` to make a copy of a DataFrame. To access a specific column from a DataFrame, use a pair of square brackets and place the name of the column as a string inside the brackets.

</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Use a pair of square brackets to create a new column in a DataFrame. The columns in DataFrames are series objects, which support elementwise operations such as multiplication and division. Be sure the column names referenced in your code match the spelling of what's in the DataFrame.
</details>

Now, to prepare for the Naive Bayes model that you will build in a later lab, clean the extracted data and ensure ensure it is concise. Naive Bayes involves an assumption that features are independent of each other given the class. In order to satisfy that criteria, if certain features are aggregated to yield new features, it may be necessary to remove those original features. Therefore, drop the columns that were used to extract new features.

**Note:** There are other types of models that do not involve independence assumptions, so this would not be required in those instances. In fact, keeping the original features may be beneficial.

In [38]:
# Remove any columns from `extracted_data` that are no longer needed.
### YOUR CODE HERE ###
filtered_df = extracted_data.drop(['pts', 'ast', 'reb', 'stl', 'blk', 'tov'], axis=1)

# Display the first few rows of `extracted_data` to ensure that column drops took place.
### YOUR CODE HERE ###
filtered_df.head()


gp   min  fga    fg  3pa    3p  fta    ft  target_5yrs  pts_per_min  \
0  36  27.4  7.6  34.7  2.1  25.0  2.3  69.9            0     0.270073   
1  35  26.9  6.7  29.6  2.8  23.5  3.4  76.5            0     0.267658   
2  74  15.3  4.7  42.2  1.7  24.4  1.3  67.0            0     0.339869   
3  58  11.6  5.5  42.6  0.5  22.6  1.3  68.9            1     0.491379   
4  48  11.5  3.0  52.4  0.1   0.0  1.9  67.4            1     0.391304   

   ast_per_min  reb_per_min  blk_per_min  stl_per_min  tov_per_min  
0     0.069343     0.149635     0.014599     0.014599     0.047445  
1     0.137546     0.089219     0.018587     0.040892     0.059480  
2     0.065359     0.143791     0.019608     0.032680     0.065359  
3     0.068966     0.163793     0.008621     0.051724     0.086207  
4     0.026087     0.217391     0.034783     0.026087     0.069565

## NOTE: 
After checking the exemplar, I disagree that columns like fga and fg are redundant. The exemplar dropped fga (field goal attempts) and kept fg (field goal percentage). I think the fga column is still very important because a player may have an extremely high or low field goal percentage due to having few field goal attempts. For example, a player may have a 100% field goal percentage on 1 field goal attempt. This could make a player look artificially good even if they do not play much (low number of gp or min, which could actually be an indication that they are a bad player). The more field goal attempts a player has, the more their fg percentage will stabilize though.

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the materials about feature extraction.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

There are functions in the `pandas` library that remove specific columns from a DataFrame and that display the first few rows of a DataFrame.
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Use the `drop()` function and pass in a list of the names of the columns you want to remove. By default, calling this function will result in a new DataFrame that reflects the changes you made. The original DataFrame is not automatically altered. You can reassign `extracted_data` to the result, in order to update it. 

Use the `head()` function to display the first few rows of a DataFrame.
</details>

Next, export the extracted data as a new .csv file. You will use this in a later lab. 

In [39]:
# Export the extracted data.
### YOUR CODE HERE ###
filtered_df.to_csv('my_extracted_nba_players_data.csv')

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

There is a function in the `pandas` library that exports a DataFrame as a .csv file. 
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Use the `to_csv()` function to export the DataFrame as a .csv file. 
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Call the `to_csv()` function on `extracted_data`, and pass in the name that you want to give to the resulting .csv file. Specify the file name as a string and in the file name. Make sure to include `.csv` as the file extension. Also, pass in the parameter `index` set to `0`, so that when the export occurs, the row indices from the DataFrame are not treated as an additional column in the resulting file. 
</details>

## **Considerations**


**What are some key takeaways that you learned during this lab? Consider the process you followed and what tasks were performed during each step, as well as important priorities when training data.**

I learned how to select and extract important features prior to training a model.

**What summary would you provide to stakeholders? Consider key attributes to be shared from the data, as well as upcoming project plans.**

I would share the features selected and extracted from the model in order to get feedback on additional features to consider prior to modeling, especially from stakeholders that have many years of domain experience (i.e., experience with performance in professional basketball leagues).

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.